### PR TITLE
statemanager: adding state file unit tests for untested features

### PR DESCRIPTION
**Pull request by fenxiong**
### Summary

Adding state file unit tests for fields related to container health check, private registry and task networking, which are not included in current state file unit tests. For each feature, the test uses the state file version when the new fields are added.

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes:  yes

### Description for the changelog

Not needed.

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/statemanager/state_manager_test.go [hunk 1]
```diff
@@ -128,3 +128,91 @@ func TestLoadsV13DataCorrectly(t *testing.T) {
...
+
+	assert.Equal(t, "state-file", cluster)
+	assert.EqualValues(t, 0, sequenceNumber)
+
+	tasks, err := taskEngine.ListTasks()
+	assert.NoError(t, err)
+	assert.Equal(t, 1, len(tasks))
+
+	task := tasks[0]
+	assert.Equal(t, "arn:aws:ecs:us-west-2:897977802564:task/e4e6c98c-aa44-4146-baf9-431b04c0d162", task.Arn)
```
[the remark this answers is not in the document]
Comment (fenxiong) [reply agent/statemanager/state_manager_test.go]: sure, i will change it to a dummy number

File: agent/statemanager/testdata/v11/task-networking/ecs_agent_data.json [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+               "expiresAt":"2018-10-04T17:37:46.286823554Z"
+            }
+         ],
+         "IPToTask":{  
+            "169.254.172.2":"arn:aws:ecs:us-west-2:897977802564:task/fad405be-8705-4175-877b-db50109a15f2"
+         }
+      }
+   },
+   "Version":11
+}
```
[the remark this answers is not in the document]
Comment (fenxiong) [reply agent/statemanager/testdata/v11/task-networking/ecs_agent_data.json]: will add

**Comment by sharanyad** [comment]
Why not have a single state file and test loads for all features in a single test?

File: agent/statemanager/state_manager_test.go [hunk 3]
```diff
@@ -128,3 +128,91 @@ func TestLoadsV13DataCorrectly(t *testing.T) {
 	assert.Equal(t, deadTask.GetKnownStatusTime(), expected)
 
 }
+
+// verify that the state manager correctly loads container health check related fields in state file
+func TestLoadsDataForContainerHealthCheckTask(t *testing.T) {
+	cleanup, err := setupWindowsTest(filepath.Join(".", "testdata", "v10", "container-health-check", "ecs_agent_data.json"))
```
Comment (sharanyad) [inline agent/statemanager/state_manager_test.go]: Should this be in a windows test file?

**Comment by fenxiong** [comment]
@sharanyad 
task networking only works for linux, but the other two work for both. if we need to separate them anyway, i think we can just have three tests, one for each?

File: agent/statemanager/state_manager_test.go (hunk 3 above)
Comment (fenxiong) [reply agent/statemanager/state_manager_test.go]: tests in this file will be run on both windows and linux, and this function is doing something that's additionally required on windows, so i think it needs to be here..

**Comment by yunhee-l** [comment]
I think there's value in keeping test cases smaller/specific to feature. 
I doubt we'd want to add multiple tests like this for every new version of state files, but if the state file change does not impact certain features, having separated out test cases will allow us to update the it only when the feature is changed.

**Comment by sharanyad** [comment]
I'm thinking in terms of time taken for running these unit tests. Currently, we have a timeout of 25s(which is reasonable, since unit tests are supposed to be quick). Since we load files from disk, for each of these tests, this will increase the testing time.

**Comment by yunhee-l** [comment]
that's a valid concern. @fenxiong  can you do some kind of benchmarking on how much longer the unit tests take with more file io-based tests?

**Comment by fenxiong** [comment]
@yunhee-l @sharanyad Tested with normal settings for our test facilities, sample results:
(1) Linux:
```
GOCACHE=off go test -run TestLoadsDataForPrivateRegistryTask -tags unit ./agent/statemanager/
ok  	github.com/aws/amazon-ecs-agent/agent/statemanager	0.006s

GOCACHE=off go test -run TestLoadsDataForPrivateRegistryTask -tags unit -count 1000 ./agent/statemanager/
ok  	github.com/aws/amazon-ecs-agent/agent/statemanager	1.007s
```
(2) Windows:
```
go test -run TestLoadsDataForPrivateRegistryTask -tags unit -count 100 ./agent/statemanager/
ok      github.com/aws/amazon-ecs-agent/agent/statemanager      0.262s
```

So basically on linux a state file unit test takes about 0.001s, and on windows each takes about 0.1-0.3s. Currently the unit tests takes a few seconds in total which is far from the limit.. so i think we are probably fine regards to test time limit?

**Pull request merged**